engine/src/snapshot_api/rocks_snapshot.rs

```diff
@@ -0,0 +1,382 @@
+                .try_into()
+                .unwrap_or_else(|_e| unreachable!("infallible"));
+            let meta = bincode::deserialize(&meta_bytes)
+                .map_err(|e| io::Error::new(ErrorKind::Other, e))?;
+
+            self.apply_snap_meta(meta);
+            let mut data = Vec::new();
+            data.extend(meta_len_bytes);
+            data.extend(meta_bytes);
+            *self.meta.data.get_mut() = data;
+            self.meta.is_current = false;
+            written_bytes = written_bytes
+                .overflow_add(meta_len.numeric_cast())
+                .overflow_add(8);
+            next_buf = &next_buf[meta_len.overflow_add(8).numeric_cast()..];
+        }
+
+        while self.snap_file_idx < self.snap_files.len() {
+            let snap_file = &mut self.snap_files[self.snap_file_idx];
+            assert!(snap_file.size != 0);
+            let left = snap_file.remain_size().numeric_cast();
+            let (write_len, switch, finished) = match next_buf.len().cmp(&left) {
+                Ordering::Greater => (left, true, false),
+                Ordering::Equal => (left, true, true),
+                Ordering::Less => (next_buf.len(), false, true),
+            };
+            snap_file.written_size = snap_file
+                .written_size
+                .overflow_add(write_len.numeric_cast());
+            written_bytes = written_bytes.overflow_add(write_len);
+
+            let buffer = &next_buf[0..write_len];
+
+            let f = if let Some(ref mut f) = self.current_file {
+                f
+            } else {
+                let path = self.current_file_path(true);
+                let writer = File::create(path).await?;
+                self.current_file = Some(writer);
+                self.current_file
+                    .as_mut()
+                    .unwrap_or_else(|| unreachable!("current_file must be `Some` here"))
+            };
+            f.write_all(buffer).await?;
+
+            if switch {
+                next_buf = &next_buf[write_len..];
+                let old = self.current_file.take();
+                if let Some(mut old_f) = old {
+                    old_f.flush().await?;
+                    let path = self.current_file_path(false);
+                    let tmp_path = self.current_file_path(true);
+                    fs::rename(tmp_path, path)?;
+                }
+                self.snap_file_idx = self.snap_file_idx.overflow_add(1);
+            }
+            if finished {
+                break;
+            }
+        }
+        Ok(written_bytes)
+    }
+}
+
+#[async_trait::async_trait]
+impl SnapshotApi for RocksSnapshot {
+    #[inline]
+    fn size(&self) -> u64 {
+        self.snap_files
+            .iter()
+            .map(|f| f.size)
+            .sum::<u64>() // sst files size
+            .overflow_add(self.meta.len().numeric_cast()) // meta size
+    }
+
+    #[inline]
+    #[allow(clippy::indexing_slicing)] // safe indexing
```

Comment:
Please explain why it's safe. For instance, the return value of `read` function is not larger than the input buf's length.

Reply:
And we should also try to put the constraint  in the `read` function. Do the same thing for the `write` function.

Reply:
Fixed